pkg/container/kubernetes_runner.go

```diff
@@ -0,0 +1,436 @@
+		logger:    logger,
+	}
+	if err := envconfig.Process("", runner); err != nil {
+		runner.logger.Warnf("failed to instantiate kubernetes runner config from environment, falling back to defaults: %v")
+	}
+
+	repository, err := name.NewRepository(runner.Repo)
+	if err != nil {
+		fallbackRepo := "ttl.sh/melange"
+		runner.logger.Warnf("failed to parse reference %s: %v, defaulting to %s", runner.Repo, err, fallbackRepo)
+		repository, _ = name.NewRepository("ttl.sh/melange")
+	}
+	runner.repository = repository
+
+	return runner
+}
+
+// Name implements Runner
+func (*k8s) Name() string {
+	return KubernetesName
+}
+
+// Run implements Runner
+func (k *k8s) Run(cfg *Config, cmd ...string) error {
+	if cfg.PodID == "" {
+		return fmt.Errorf("pod isn't running")
+	}
+
+	c, kcfg, err := k.client()
+	if err != nil {
+		return err
+	}
+
+	k.logger.Printf("Executing: %s", cmd)
+	req := c.CoreV1().
+		RESTClient().
+		Post().
+		Resource("pods").
+		Name(cfg.PodID).
+		Namespace(k.Namespace).
+		SubResource("exec").
+		VersionedParams(&corev1.PodExecOptions{
+			Command: cmd,
+			Stdin:   false,
+			Stdout:  true,
+			Stderr:  true,
+			TTY:     false,
+		}, scheme.ParameterCodec)
+
+	exec, err := remotecommand.NewSPDYExecutor(kcfg, "POST", req.URL())
+	if err != nil {
+		return fmt.Errorf("failed to create remote command executor: %v", err)
+	}
+
+	stdoutPipeR, stdoutPipeW, err := os.Pipe()
+	if err != nil {
+		return err
+	}
+	stderrPipeR, stderrPipeW, err := os.Pipe()
+	if err != nil {
+		return err
+	}
+	finishStdout := make(chan struct{})
+	finishStderr := make(chan struct{})
+
+	go monitorPipe(cfg.Logger, log.InfoLevel, stdoutPipeR, finishStdout)
+	go monitorPipe(cfg.Logger, log.WarnLevel, stderrPipeR, finishStderr)
+
+	if err := exec.Stream(remotecommand.StreamOptions{
+		Stdout: stdoutPipeW,
+		Stderr: stderrPipeW,
+	}); err != nil {
+		return fmt.Errorf("failed to stream remote command: %v", err)
+	}
+
+	stdoutPipeW.Close()
+	stderrPipeW.Close()
+
+	<-finishStdout
+	<-finishStderr
+	return nil
+}
+
+// StartPod implements Runner
+func (k *k8s) StartPod(cfg *Config) error {
+	if cfg.PodID != "" {
+		return fmt.Errorf("pod already running: %s", cfg.PodID)
+	}
+
+	builder := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: fmt.Sprintf("%s%s-", k.Prefix, cfg.PackageName),
+			Namespace:    k.Namespace,
+			Labels: map[string]string{
+				"app.kubernetes.io/component":    cfg.PackageName,
+				"melange.chainguard.dev/package": cfg.PackageName,
+				"melange.chainguard.dev/arch":    cfg.Arch.String(),
+			},
+		},
+		Spec: corev1.PodSpec{
+			Containers: []corev1.Container{{
+				Name:    "melange",
+				Image:   cfg.ImgRef,
+				Command: []string{"sleep", "infinity"},
+				Resources: corev1.ResourceRequirements{
+					Requests: corev1.ResourceList{
+						corev1.ResourceCPU:    resource.Quantity(k.CPU),
+						corev1.ResourceMemory: resource.Quantity(k.Memory),
+					},
+				},
+			}},
+		},
+	}
+
+	if cfg.Arch.ToAPK() == "aarch64" {
```

Comment:
Why not set `kubernetes.io/arch` unconditionally?  `kubernetes.io/arch: amd64` is a no-op.

Something like:

```go
builder.Spec.NodeSelector = map[string]string{
  "kubernetes.io/arch": cfg.Arch.String(),
  "cloud.google.com/compute-class": "Scale-Out",
}
```

cfg.Arch default string representation is always the same as `runtime.GOARCH`, which matches the kubernetes selector.  This futureproofs us for when we eventually add `riscv64`, `ppc64le` or `s390x`.